jupyterhub/utils.py

```diff
@@ -48,29 +49,62 @@ def can_connect(ip, port):
     else:
         return True
 
-# exponential falloff factors:
-# start at 100ms, falloff by 2x
-# never longer than 5s
-DT_MIN = 0.1
-DT_SCALE = 2
-DT_MAX = 5
+@gen.coroutine
+def exponential_backoff(
+        pass_func,
+        fail_message,
+        start_wait=0.1,
+        scale_factor=2,
+        max_wait=5,
+        timeout=10,
+        *args, **kwargs):
+    """
+    Exponentially backoff until pass_func is true.
+
+    This function will wait with exponential backoff + random jitter for as
+    many iterations as needed, with maximum timeout timeout. If pass_func is
+    still returning false at the end of timeout, a TimeoutError will be raised.
+
+    It'll start waiting at start_wait, scaling up by continuously multiplying itself
+    by scale_factor until pass_func returns true. It'll never wait for more than
+    max_wait seconds per iteration.
+
+    *args and **kwargs are passed to pass_func. pass_func maybe a future, although
+    that is not entirely recommended.
+
+    It'll return the value of pass_func when it's truthy!
+    """
+    loop = ioloop.IOLoop.current()
+    start_tic = loop.time()
+    dt = start_wait
+    while True:
+        if (loop.time() - start_tic) > timeout:
+            # We time out!
+            break
+        ret = yield gen.maybe_future(pass_func(*args, **kwargs))
+        # Truthy!
+        if ret:
+            return ret
+        else:
+            yield gen.sleep(dt)
+        # Add some random jitter to improve performance
+        # This makes sure that we don't overload any single iteration
+        # of the tornado loop with too many things
+        # See https://www.awsarchitectureblog.com/2015/03/backoff.html
+        # for a good example of why and how this helps. We're using their
+        # full Jitter implementation equivalent.
+        dt = min(max_wait, random.uniform(0, dt * scale_factor))
```

Comment:
Another point: probably want to introduce jitter on the first `dt` before `gen.sleep`, right? Rather than only after the second iteration? I think the dt calculation also probably makes more sense before gen.sleep. Maybe something like this:

```python
if ret:
    return ret
deadline = timeout - (loop.time() - start_tic)
if deadline < 0:
    # timeout exceeded
    break
dt = min(max_wait, deadline, random.uniform(0, dt * scale_factor))
yield gen.sleep(dt)
```